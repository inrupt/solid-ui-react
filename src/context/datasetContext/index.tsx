//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import React, { createContext, ReactElement, useState, useEffect } from "react";
import {
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import useDataset from "../../hooks/useDataset";

export interface IDatasetContext {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo) | undefined;
  setDataset: (solidDataset: SolidDataset) => void;
}

const DatasetContext = createContext<IDatasetContext>({
  solidDataset: undefined,
  setDataset: () => {},
});

export default DatasetContext;

export interface IDatasetProvider {
  children: React.ReactNode;
  /**
   * @deprecated `loading` is deprecated. It can still be used but may be removed in a future major release.`loadingComponent` should be used instead.
   */
  loading?: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
  solidDataset?: SolidDataset | (SolidDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireDatasetOrDatasetUrl =
  | RequireProperty<IDatasetProvider, "solidDataset">
  | RequireProperty<IDatasetProvider, "datasetUrl">;

/**
 * Used to provide a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) to child components through context, as used by various provided components and the useDataset hook.
 */
export const DatasetProvider = ({
  children,
  onError,
  solidDataset: propDataset,
  datasetUrl,
  loading,
  loadingComponent: LoadingComponent,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const { dataset, error } = useDataset(datasetUrl);

  if (error && onError) {
    onError(error);
  }

  const datasetToUse = propDataset ?? dataset;

  // Provide a setDataset function so that child components can update.
  const [stateDataset, setDataset] = useState<
    SolidDataset | (SolidDataset & WithResourceInfo) | undefined
  >(datasetToUse);

  // If the dataset is asynchronously loaded, make sure to set the new state value.
  useEffect(() => {
    setDataset(datasetToUse);
  }, [datasetToUse]);

  let loader: JSX.Element | React.ReactNode | null = (LoadingComponent && (
    <LoadingComponent />
  )) ||
    loading || <span>Fetching data...</span>;

  if (LoadingComponent === null) {
    loader = null;
  }

  return (
    <DatasetContext.Provider value={{ solidDataset: stateDataset, setDataset }}>
      {stateDataset ? children : loader}
    </DatasetContext.Provider>
  );
};
