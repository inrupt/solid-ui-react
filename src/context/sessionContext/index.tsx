/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  SetStateAction,
  Dispatch,
  useEffect,
  ReactNode,
} from "react";
import auth from "solid-auth-client";

export interface ISession {
  webId: string;
}

interface ISessionContext {
  session?: ISession | undefined;
  setSession?: Dispatch<SetStateAction<ISession>> | any;
  sessionRequestInProgress?: boolean;
  setSessionRequestInProgress?: Dispatch<SetStateAction<boolean>> | any;
}

export const SessionContext = createContext<ISessionContext>({});

interface ISessionProvider {
  children: ReactNode;
  session?: ISession;
  sessionRequestInProgress?: boolean;
}

export const SessionProvider = ({
  session: initialSession,
  sessionRequestInProgress: initialSessionRequestInProgress,
  children,
}: ISessionProvider): ReactElement => {
  const [session, setSession] = useState<ISession | undefined>(initialSession);
  const [sessionRequestInProgress, setSessionRequestInProgress] = useState(
    initialSessionRequestInProgress
  );

  useEffect(() => {
    auth.trackSession(setSession).catch((error) => {
      throw error;
    });

    return function cleanup() {
      auth.stopTrackSession(setSession);
    };
  }, [setSession]);

  return (
    <SessionContext.Provider
      value={{
        session,
        setSession,
        sessionRequestInProgress,
        setSessionRequestInProgress,
      }}
    >
      {children}
    </SessionContext.Provider>
  );
};

SessionProvider.defaultProps = {
  session: undefined,
  sessionRequestInProgress: false,
};
