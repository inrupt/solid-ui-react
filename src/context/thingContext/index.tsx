/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  useEffect,
  useContext,
} from "react";
import {
  Thing,
  LitDataset,
  WithResourceInfo,
  UrlString,
  getThingOne,
} from "@inrupt/solid-client";
import { DatasetContext } from "../datasetContext";

interface IThingContext {
  thing: Thing | undefined;
}

export const ThingContext = createContext<IThingContext>({
  thing: undefined,
});

interface IThingProvider {
  children: React.ReactNode;
  thing?: Thing;
  thingUrl?: UrlString | string;
  dataset?: LitDataset | (LitDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

type RequireProperty<T, Prop extends keyof T> = T & { [key in Prop]-?: T[key] };
type RequireDatasetOrDatasetUrl =
  | RequireProperty<IThingProvider, "thing">
  | RequireProperty<IThingProvider, "thingUrl">;

export const ThingProvider = ({
  children,
  thing,
  thingUrl,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const datasetContext = useContext(DatasetContext);
  const { dataset } = datasetContext;
  const [litThing, setLitThing] = useState<Thing>();

  useEffect(() => {
    if (dataset && thingUrl) {
      setLitThing(getThingOne(dataset, thingUrl));
    }
  }, [dataset, thingUrl]);

  if (thing) {
    return (
      <ThingContext.Provider
        value={{
          thing,
        }}
      >
        {children}
      </ThingContext.Provider>
    );
  }

  return (
    <ThingContext.Provider
      value={{
        thing: litThing,
      }}
    >
      {children}
    </ThingContext.Provider>
  );
};
