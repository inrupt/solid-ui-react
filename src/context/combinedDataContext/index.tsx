//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import type { ReactElement } from "react";
import React from "react";

import type {
  Thing,
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import { DatasetProvider } from "../datasetContext";
import { ThingProvider } from "../thingContext";

export type Props = {
  children: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
};

export type ThingOrThingUrl = {
  thing?: Thing;
  thingUrl?: UrlString | string;
};

export interface Dataset extends Props {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo);
}

export interface DatasetUrl extends Props {
  datasetUrl: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireThingOrThingUrl =
  | RequireProperty<ThingOrThingUrl, "thing">
  | RequireProperty<ThingOrThingUrl, "thingUrl">;

// TODO: naming of this provider TBC
function CombinedDataProvider(
  props: Dataset & RequireThingOrThingUrl,
): ReactElement;

function CombinedDataProvider(
  props: DatasetUrl & RequireThingOrThingUrl,
): ReactElement;

/**
 * Used to provide both a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) and [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) to child components through context, as used by various provided components and the useDataset and useThing hooks.
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function CombinedDataProvider(props: any): ReactElement {
  const {
    children,
    solidDataset,
    datasetUrl,
    thing,
    thingUrl,
    onError,
    loadingComponent,
  } = props;

  return (
    <DatasetProvider
      onError={onError}
      solidDataset={solidDataset}
      datasetUrl={datasetUrl}
      loadingComponent={loadingComponent}
    >
      <ThingProvider thing={thing} thingUrl={thingUrl}>
        {children}
      </ThingProvider>
    </DatasetProvider>
  );
}

export default CombinedDataProvider;
