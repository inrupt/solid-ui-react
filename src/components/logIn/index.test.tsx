/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import * as React from "react";
import {
  render,
  RenderResult,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import auth from "solid-auth-client";
import LoginButton from "./index";
import { SessionContext } from "../../context/sessionContext";

jest.mock("solid-auth-client");
let documentBody: RenderResult;
const onLogin = jest.fn();
const onError = jest.fn();
const setSessionRequestInProgress = jest.fn();

describe("<LoginButton /> component snapshot test", () => {
  beforeEach(() => {
    documentBody = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton
          popupUrl="./popup.html"
          onLogin={onLogin}
          onError={onError}
        />
      </SessionContext.Provider>
    );
  });
  it("matches snapshot", () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});

describe("<LoginButton /> component visual testing", () => {
  it("Renders child element", () => {
    const { getByText } = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton
          popupUrl="./popup.html"
          onLogin={onLogin}
          onError={onError}
        >
          <div>Custom child element</div>
        </LoginButton>
      </SessionContext.Provider>
    );

    expect(getByText("Custom child element")).toBeTruthy();
  });
});

describe("<LoginButton /> component functional testing", () => {
  it("fires the onClick function and calls OnLogin", async () => {
    (auth.popupLogin as jest.Mock).mockImplementationOnce(() =>
      Promise.resolve()
    );

    const { getByText } = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton
          popupUrl="./popup.html"
          onLogin={onLogin}
          onError={onError}
        />
      </SessionContext.Provider>
    );

    fireEvent.click(getByText("Log In"));
    expect(setSessionRequestInProgress).toHaveBeenCalledTimes(1);
    expect(auth.popupLogin).toHaveBeenCalled();

    await waitFor(() => expect(onLogin).toHaveBeenCalledTimes(1));
  });

  it("fires the onClick function and doesn't call OnLogin if it wasn't provided", async () => {
    (auth.popupLogin as jest.Mock).mockImplementationOnce(() =>
      Promise.resolve()
    );

    const { getByText } = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton popupUrl="./popup.html" onError={onError} />
      </SessionContext.Provider>
    );

    fireEvent.click(getByText("Log In"));
    await waitFor(() => expect(onLogin).toHaveBeenCalledTimes(0));
  });

  it("fires the onClick function and calls OnError", async () => {
    (auth.popupLogin as jest.Mock).mockImplementationOnce(() =>
      Promise.reject()
    );
    const { getByText } = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton
          popupUrl="./popup.html"
          onLogin={onLogin}
          onError={onError}
        />
      </SessionContext.Provider>
    );
    fireEvent.click(getByText("Log In"));
    expect(auth.popupLogin).toHaveBeenCalled();
    await waitFor(() => expect(onError).toHaveBeenCalledTimes(1));
  });

  it("fires the onClick function and doesn't call OnError if it wasn't provided", async () => {
    (auth.popupLogin as jest.Mock).mockImplementationOnce(() =>
      Promise.reject()
    );
    const { getByText } = render(
      <SessionContext.Provider value={{ setSessionRequestInProgress }}>
        <LoginButton popupUrl="./popup.html" onLogin={onLogin} />
      </SessionContext.Provider>
    );
    fireEvent.click(getByText("Log In"));
    await waitFor(() => expect(onError).toHaveBeenCalledTimes(0));
  });
});
