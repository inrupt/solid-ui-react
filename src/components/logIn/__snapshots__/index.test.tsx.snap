// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginButton /> component snapshot test matches snapshot 1`] = `
<body>
  <div>
    <button
      type="button"
    >
      Log In
    </button>
  </div>
</body>
`;
